src/Client/PacketReceiver.h

```diff
@@ -0,0 +1,225 @@
+#pragma once
+
+#if defined(OS_LINUX)
+
+#include <variant>
+
+#include <Client/IConnections.h>
+#include <Common/FiberStack.h>
+#include <Common/Fiber.h>
+#include <Common/Epoll.h>
+#include <Common/TimerDescriptor.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int CANNOT_OPEN_FILE;
+    extern const int CANNOT_READ_FROM_SOCKET;
+}
+
+/// Class for nonblocking packet receiving. It runs connection->receivePacket
+/// in fiber and sets special read callback which is called when
+/// reading from socket blocks. When read callback is called,
+/// socket and receive timeout are added in epoll and execution returns to the main program.
+/// So, you can poll this epoll file descriptor to determine when to resume
+/// packet receiving.
+class PacketReceiver
+{
+public:
+    explicit PacketReceiver(Connection * connection_) : connection(connection_)
+    {
+        epoll.add(receive_timeout.getDescriptor());
+        epoll.add(connection->getSocket()->impl()->sockfd());
+
+        if (-1 == pipe2(pipe_fd, O_NONBLOCK))
+            throwFromErrno("Cannot create pipe", ErrorCodes::CANNOT_OPEN_FILE);
+        epoll.add(pipe_fd[0]);
+
+        fiber = boost::context::fiber(std::allocator_arg_t(), fiber_stack, Routine{*this});
+    }
+
+    ~PacketReceiver()
+    {
+        close(pipe_fd[0]);
+        close(pipe_fd[1]);
+    }
+
+    /// Resume packet receiving.
+    std::variant<int, Packet, Poco::Timespan> resume()
+    {
+        /// If there is no pending data, check receive timeout.
+        if (!connection->hasReadPendingData() && !checkReceiveTimeout())
+        {
+            /// Receive timeout expired.
+            return Poco::Timespan();
+        }
+
+        /// Resume fiber.
+        fiber = std::move(fiber).resume();
+        if (exception)
+            std::rethrow_exception(std::move(exception));
+
+        if (is_read_in_process)
+            return epoll.getFileDescriptor();
+
+        /// Write something in pipe when buffer has pending data, because
+        /// in this case socket won't be ready in epoll but we need to tell
+        /// outside that there is more data in buffer.
+        if (connection->hasReadPendingData())
+        {
+            uint64_t buf = 0;
+            while (-1 == write(pipe_fd[1], &buf, sizeof(buf)))
+            {
+                if (errno == EAGAIN)
+                    break;
+
+                if (errno != EINTR)
+                    throwFromErrno("Cannot write to pipe", ErrorCodes::CANNOT_READ_FROM_SOCKET);
+            }
+        }
```

Comment:
That's hidden logic, and I don't understand why it is needed.
When `resume` returns packet, client should process packer and call `resume` again (if it is needed).
I think it would be more logical just to add another one queue with replicas which returned packet and need to be resumed.